examples/src/streaming/client.rs

```diff
@@ -2,28 +2,84 @@ pub mod pb {
     tonic::include_proto!("grpc.examples.echo");
 }
 
+use futures::stream::Stream;
 use pb::{echo_client::EchoClient, EchoRequest};
+use std::time::Duration;
+use tokio_stream::StreamExt;
+use tonic::transport::Channel;
 
-#[tokio::main]
-async fn main() -> Result<(), Box<dyn std::error::Error>> {
-    let mut client = EchoClient::connect("http://[::1]:50051").await.unwrap();
+fn echo_requests_iter() -> impl Stream<Item = EchoRequest> {
+    tokio_stream::iter(1..usize::MAX).map(|i| EchoRequest {
+        message: format!("msg {:02}", i),
+    })
+}
 
+async fn streaming_echo(client: &mut EchoClient<Channel>, num: usize) {
     let stream = client
         .server_streaming_echo(EchoRequest {
             message: "foo".into(),
         })
         .await
+        .unwrap()
+        .into_inner();
+
+    let rev_task = tokio::spawn(async move {
+        // stream is infinite - take just 5 elements and then disconnect
+        let mut stream = stream.take(num);
+        while let Some(item) = stream.next().await {
+            println!("\trecived: {}", item.unwrap().message);
+        }
+        // stream is droped here and the disconnect info is send to server
+    });
+
+    rev_task.await.unwrap();
+}
+
+async fn bidirectional_streaming_echo(client: &mut EchoClient<Channel>, num: usize) {
+    let in_stream = echo_requests_iter().take(num);
+
+    let response = client
+        .bidirectional_streaming_echo(in_stream)
+        .await
         .unwrap();
 
-    println!("Connected...now sleeping for 2 seconds...");
+    let mut resp_stream = response.into_inner();
+
+    while let Some(recived) = resp_stream.next().await {
+        let recived = recived.unwrap();
+        println!("\trecived message: `{}`", recived.message);
+    }
+}
+
+async fn bidirectional_streaming_echo_throttle(client: &mut EchoClient<Channel>, dur: Duration) {
```

Comment:
Not following why we want this example as well?

Reply:
I wanted to show difference between graceful stream end the one when client disconnected for example by broken connection or just crush. This 2 functions could be unified to one. 

Here is the output from server part.
```
EchoServer::server_streaming_echo
	client connected from: Some([::1]:36750)
	client disconnected
EchoServer::Bidirectional_streaming_echo
	stream ended
EchoServer::Bidirectional_streaming_echo
	client disconnected: broken pipe
	stream ended
```

Reply:
In other words the difference is on the server side. On the client side I just needed a function that sends request slowly enough that you can write what is printed. I added additional comments to clarify that.

Should I unify this to `bidirectional_streaming_echo(client &mut EchoClient<Channel>, num_reqs: usize, throttle: Duration)`?